### PR TITLE
[java] DoNotUseThreads should not warn on Runnable #1627

### LINKED ISSUE #1627: [java] DoNotUseThreads should not warn on Runnable
**Affects PMD Version:** 

6.11.0

**Rule:**

[DoNotUseThreads](https://pmd.github.io/pmd-6.11.0/pmd_rules_java_multithreading.html#donotusethreads)

**Description:**

I argue that, since Java 8, Runnables are the de-facto standard way to model `()->void` functions. I am aware that the Javadoc says

>The Runnable interface should be implemented by any class whose instances are intended to be executed by a thread

yet, there is no substitute in `java.util.function`, and until something else is in place, in several cases a delayed computation is modelled through `Runnable`, generating a number of false positives.

I'd like to open a discussion on whether the rule should be changed to take into account only `Thread` (and, possibily, `Executors` and `ExecutorService`).

**Code Sample demonstrating the issue:**

Snippet from Protelis/Protelis

```java
    @Override
    public void eval(final ExecutionContext context) {
        final AnnotatedTree<?> initBranch = getBranch(INIT);
        initBranch.evalInNewStackFrame(context, INIT);
        final S localValue = ensureType(isErased() ? initBranch.getAnnotation() : getSuperscript());
        final AnnotatedTree<?> body = getBranch(BODY);
        ifPresent(localName, it -> context.putVariable(it, localValue, true));
        ifPresent(fieldName, it -> context.putVariable(it, context.buildFieldDeferred(i -> i, localValue, body::getAnnotation), true));
        context.newCallStackFrame(BODY);
        final Runnable yieldEvaluation = () -> ifPresent(yield, it -> it.evalInNewStackFrame(context, YIELD)); // Warning, but this is just a delayed computation, no threads involved...
        if (body instanceof All) {
            final All multilineBody = (All) body;
            multilineBody.forEachWithIndex((i, b) -> {
                context.newCallStackFrame(i.byteValue());
                b.eval(context);
                multilineBody.setAnnotation(b.getAnnotation());
            });
            yieldEvaluation.run();
            multilineBody.forEach(it -> context.returnFromCallFrame());
        } else {
            body.eval(context);
            yieldEvaluation.run();
        }
        context.returnFromCallFrame();
        final S result = ensureType(body.getAnnotation());
        setSuperscript(result);
        setAnnotation(yield.isPresent() ? yield.get().getAnnotation() : (T) result);
    }
```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

### PR DESCRIPTION
**Pull request by mclay**
Before submitting a PR, please check that:
 - [x] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [x] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

**PR Description:**
Fixes #1627

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 110 new violations, 0 new errors and 0 new configuration errors,<br />
removes 694 violations, 2 errors and 2 configuration errors.<br />
<a href="https://chunk.io/pmd/e3766a1ea24340bba5147d291b429675/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/multithreading/xml/DoNotUseThreads.xml [hunk 1]
```diff
@@ -29,32 +29,80 @@ extending threads is not allowed
 			]]>
         </code>
     </test-code>
-        <test-code>
+    <test-code>
         <description><![CDATA[
-implementing runnabel is not allowed
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/multithreading/xml/DoNotUseThreads.xml]: I think, we should add a extra test case, that explicitly allows that runnable can be used (so keep this test case, but change the expected problems to 0) - referencing the github issue #1627 where is has been discussed. Without this test case, we don't have prove that this issue is fixed indeed (and stays fixed in the future).

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/multithreading/xml/DoNotUseThreads.xml [hunk 2]
```diff
@@ -29,32 +29,80 @@ extending threads is not allowed
...
+            <![CDATA[
+            import java.util.concurrent.*;
+
+            public class PMDTestExecutorService extends AbstractExecutorService {}
+         ]]>
+        </code>
+    </test-code>
+    <test-code>
+        <description><![CDATA[
+        Using ExecutorService is not allowed
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/multithreading/xml/DoNotUseThreads.xml]: FYI - the CDATA sections are here unnecessary, just write it without:

`<description>Using ExecutorService is not allowed</description>`

**Comment by mclay** [comment]
@adangel thanks for the review and feedback. I have incorporated all your suggestions.

**Pull request merged**